exec: integer division causes a panic

Vectorized execution panics when you divide an int column by an int constant, presumably because it thinks the result should be an integer but it should actually be a decimal.
```
root@127.0.0.1:61735/defaultdb> set experimental_vectorize=on;
SET

Time: 281µs

root@127.0.0.1:61735/defaultdb> create table t (x int);
CREATE TABLE

Time: 2.721ms

root@127.0.0.1:61735/defaultdb> insert into t values (1);
INSERT 1

Time: 2.863ms

root@127.0.0.1:61735/defaultdb> select x/2 from t;
E190710 20:17:30.876509 724 sql/conn_executor.go:713  [n1,client=127.0.0.1:61737,user=root] a SQL panic has occurred while executing "SELECT x / 2 FROM t": interface conversion: coldata.column is []int64, not []apd.Decimal
E190710 20:17:30.876569 724 util/log/crash_reporting.go:207  [n1,client=127.0.0.1:61737,user=root] a panic has occurred!
panic: interface conversion: coldata.column is []int64, not []apd.Decimal [recovered]
	panic: panic while executing 1 statements: SELECT _ / _ FROM _; caused by interface conversion: coldata.column is []int64, not []apd.Decimal

goroutine 724 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00153b600, 0x77fbde0, 0xc00158d640, 0x6bd5e80, 0xc001175710)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00153b600, 0x77fbde0, 0xc00158d640)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x6bd5e80, 0xc001175710)
	/usr/local/Cellar/go/1.12.6/libexec/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/exec/coldata.(*memColumn).Decimal(0xc001176dc0, 0x0, 0x785df60, 0xc001176dc0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/coldata/vec.go:226 +0x74
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).Next(0xc000926e00, 0x59b2eb3, 0xc0011753e0, 0xc001bc2290, 0x5a5a4de)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:166 +0xb87
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x77fbea0, 0xc001175200, 0x78079e0, 0xc000926e00, 0x77d87e0, 0xc001430000)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc000926e00, 0x77fbea0, 0xc001175200)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc0001232c0, 0x77fbea0, 0xc001175200, 0x713a620, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:654 +0x1f2
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0001e8fc0, 0xc001aca960, 0xc0014ba240, 0xc001bc28b0, 0xc001430000, 0xc00153ba28, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:295 +0x35c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc0001e8fc0, 0x77fbea0, 0xc001174c90, 0xc00153ba28, 0xc001aca960, 0xc0014ba240, 0x77fdba0, 0xc0014aad80, 0xc001430000)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:859 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc00153b600, 0x77fbea0, 0xc001174c90, 0xc00153b950, 0x3, 0xfa19d18, 0xc0007bf8c0, 0xc001bc2b01, 0x0, 0x0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:902 +0x370
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc00153b600, 0x77fbea0, 0xc001174c90, 0xc00153b950, 0xfa19d18, 0xc0007bf8c0, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:734 +0x6e5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc00153b600, 0x77fbea0, 0xc001174c90, 0x78013a0, 0xc001118740, 0xc0015531ee, 0x11, 0x0, 0x1, 0x0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:417 +0xb64
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc00153b600, 0x77fbea0, 0xc001174c90, 0x78013a0, 0xc001118740, 0xc0015531ee, 0x11, 0x0, 0x1, 0x0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:99 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc00153b600, 0x77fbde0, 0xc001176780, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1204 +0x1ab0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc00153b600, 0x77fbde0, 0xc00158d640, 0xc00010e340, 0x5400, 0x15000, 0xc00010e3d8, 0xc000087160, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1140 +0x1a3
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0001ddb00, 0x77fbde0, 0xc00158d640, 0xc00153b600, 0x5400, 0x15000, 0xc00010e3d8, 0xc000087160, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc00167c299, 0xc00162e860, 0x77fbde0, 0xc00158d640, 0xc000087160, 0xc0001ddb00, 0xc00176e000, 0x77ff920, 0xc00162e840, 0xc0015674a0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580 +0x21c
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:509 +0x17b
```

--- a/pkg/sql/distsqlrun/column_exec_setup.go
+++ b/pkg/sql/distsqlrun/column_exec_setup.go
@@ -543,6 +543,9 @@ func newColOperator(
 func planSelectionOperators(
 	ctx *tree.EvalContext, expr tree.TypedExpr, columnTypes []semtypes.T, input exec.Operator,
 ) (op exec.Operator, resultIdx int, ct []semtypes.T, err error) {
+	if err := assertHomogeneousTypes(expr); err != nil {
+		return op, resultIdx, ct, err
+	}
 	switch t := expr.(type) {
 	case *tree.IndexedVar:
 		return exec.NewBoolVecToSelOp(input, t.Idx), -1, columnTypes, nil
@@ -583,12 +586,6 @@ func planSelectionOperators(
 		if err != nil {
 			return nil, resultIdx, ct, err
 		}
-		if !ct[leftIdx].Identical(&ct[rightIdx]) {
-			err = errors.Errorf(
-				"comparison between %s and %s is unhandled", ct[leftIdx].Family(),
-				ct[rightIdx].Family())
-			return nil, resultIdx, ct, err
-		}
 		op, err := exec.GetSelectionOperator(typ, cmpOp, rightOp, leftIdx, rightIdx)
 		return op, resultIdx, ct, err
 	default:
@@ -603,6 +600,9 @@ func planSelectionOperators(
 func planProjectionOperators(
 	ctx *tree.EvalContext, expr tree.TypedExpr, columnTypes []semtypes.T, input exec.Operator,
 ) (op exec.Operator, resultIdx int, ct []semtypes.T, err error) {
+	if err := assertHomogeneousTypes(expr); err != nil {
+		return op, resultIdx, ct, err
+	}
 	resultIdx = -1
 	switch t := expr.(type) {
 	case *tree.IndexedVar:
@@ -693,16 +693,36 @@ func planProjectionExpr(
 	if err != nil {
 		return nil, resultIdx, nil, err
 	}
-	if !ct[leftIdx].Identical(&ct[rightIdx]) {
-		err = errors.Errorf(
-			"projection on %s and %s is unhandled", ct[leftIdx].Family(),
-			ct[rightIdx].Family())
-		return nil, resultIdx, ct, err
-	}
 	resultIdx = len(ct)
 	op, err = exec.GetProjectionOperator(typ, binOp, rightOp, leftIdx, rightIdx, resultIdx)
 	ct = append(ct, *typ)
 	return op, resultIdx, ct, err
+}
+
+// assertHomogeneousTypes checks that the left and right sides of an expression
+// have identical types. (Vectorized execution does not yet handle mixed types.)
+// For BinaryExprs, it also checks that the result type matches, since this is
+// not the case for certain operations like integer division.
+func assertHomogeneousTypes(expr tree.TypedExpr) error {
+	switch t := expr.(type) {
+	case *tree.BinaryExpr:
+		left := t.TypedLeft().ResolvedType()
+		right := t.TypedRight().ResolvedType()
+		result := t.ResolvedType()
+		if !left.Identical(right) {
+			return errors.Errorf("BinaryExpr on %s and %s is unhandled", left, right)
+		}
+		if !left.Identical(result) {
+			return errors.Errorf("BinaryExpr on %s with %s result is unhandled", left, result)
+		}
+	case *tree.ComparisonExpr:
+		left := t.TypedLeft().ResolvedType()
+		right := t.TypedRight().ResolvedType()
+		if !left.Identical(right) {
+			return errors.Errorf("ComparisonExpr on %s and %s is unhandled", left, right)
+		}
+	}
+	return nil
 }
 
 // wrapWithVectorizedStatsCollector creates a new exec.VectorizedStatsCollector

--- a/pkg/sql/logictest/testdata/logic_test/vectorize
+++ b/pkg/sql/logictest/testdata/logic_test/vectorize
@@ -400,3 +400,15 @@ query O
 SELECT * FROM t38754
 ----
 1
+
+# Test integer division. (Should fall back due to decimal result.)
+query T
+SELECT a/b FROM a WHERE b = 2
+----
+0.5
+
+# Test mixed types comparison. (Should also fall back.)
+query I
+SELECT b FROM a WHERE b < 0.5
+----
+0

